Doc: pci-vuart-guide (#5537)
projectacrn/acrn-hypervisor: doc/tutorials/vuart_configuration.rst

   the hypervisor is not sufficient. Currently, we recommend that you use
   the configuration in the figure 3 data flow. This may be refined in the
   future.

Use PCI-vUART
#############

PCI Interface of ACRN vUART
===========================

> When user set `vuart[0] and vuart[1] <https://projectacrn.github.io/latest/tutorials/vuart_configuration.html#introduction>`_,

Reviewer: Please use a reference to a label when cross-referencing documents/paragraphs

Author: OK, will use reference label

Index of vUART
==============

ACRN hypervisor support PCI-vUART and legacy vUART, they are all ACRN vUART. Each vUART port has its own vuart_idx.
ACRN hypervisor supports up to 8 vUART for each VM, from vuart_idx=0 to vuart_idx=7.
Suppose we use vUART0 for port with vuart_idx=0, vUART1 for vuart_idx=1, and so on. Please pay attention to these points:

* vUART0 is console port, vUART1~vUART7 are communication port.
* Each communication port must set connection to another communication vUART port of another VM
* When legacy vuart[0] is available, it is vUART0. PCI-vUART can't be vUART0 unless vuart[0] is not set.
* When legacy vuart[1] is available, it is vUART1. PCI-vUART can't be vUART1 unless vuart[1] is not set.

Setup ACRN vUART Using Configuration Tools
==========================================

When user setup ACRN VM configurations with PCI-vUART, it is better to use configuration tools. Because PCI resources are required: BDF number, address and size of mmio registers, address and size of MSIX entry tableas. They can't conflict with another PCI devices.
Futhermore, whether PCI-vUART can use vuart_idx=0 and vuart_idx=1, depend on legacy vUART settings.

> So, configuration tools will override your settings in `How to configure a console port <https://projectacrn.github.io/latest/tutorials/vuart_configuration.html#how-to-configure-a-console-port>`_ and `How to configure a communication port <https://projectacrn.github.io/latest/tutorials/vuart_configuration.html#how-to-configure-a-communication-port>`_

Reviewer: Same here, please use a :ref:`label` instruction.

Author: OK, will use reference label

      </legacy_vuart>
      <console_vuart id="0">
        <base>INVALID_PCI_BASE</base>     /* PCI-vUART0 can't be used, because vuart[0] */
      </console_vuart>
      <communication_vuart id="1">
        <base>PCI_VUART</base>            /* PCI-vUART1 is communication port, connect to vUART1 of VM1 */
        <target_vm_id>1</target_vm_id>
        <target_uart_id>1</target_uart_id>
      </communication_vuart>
   </vm>

   <vm id="1">
      <legacy_vuart id="0">
        <type>VUART_LEGACY_PIO</type>
        <base>INVALID_COM_BASE</base>     /* vuart[0] is not used */
      </legacy_vuart>
      <legacy_vuart id="1">
        <type>VUART_LEGACY_PIO</type>
        <base>INVALID_COM_BASE</base>     /* vuart[1] is not used */
      </legacy_vuart>
      <console_vuart id="0">
        <base>PCI_VUART</base>            /* PCI-vUART0 is console port */
      </console_vuart>
      <communication_vuart id="1">
        <base>PCI_VUART</base>            /* PCI-vUART1 is commincation port, connect to vUART1 of VM0 */
        <target_vm_id>0</target_vm_id>
        <target_uart_id>1</target_uart_id>
      </communication_vuart>
   </vm>

> The ACRN vUART related xml fields:

Reviewer: ```suggestion
The ACRN vUART related XML fields:
```

Author: Will fix it

>  - 'id' in <legacy_vuart>, value of vuart_idx, id=0 is for legacy vuart[0] configuration, id=1 is for vuart[1].

Reviewer: ```suggestion
 - ``id`` in <legacy_vuart>, value of vuart_idx, id=0 is for legacy vuart[0] configuration, id=1 is for vuart[1].
```

Use these double-backticks for all fields below please.

Author: OK, will use 2 backticks

 - 'type' in <legacy_vuart>, type is always VUART_LEGACY_PIO for legacy vUART.

>  - 'base' in <legacy_vuart>, is use the legacy vUART port, set COM1_BASE for vuart[0], set COM2_BASE for vuart[1].INVALID_COM_BASE means do not use the legacy vUART port.

Reviewer: ```suggestion
 - 'base' in <legacy_vuart>, is use the legacy vUART port, set COM1_BASE for vuart[0], set COM2_BASE for vuart[1]. ``INVALID_COM_BASE`` means do not use the legacy vUART port.
```

Author: OK

 - 'irq' in <legacy_vuart>, is use the legacy vUART port, set COM1_IRQ for vuart[0], set COM2_IRQ for vuart[1];
 - 'id' in <console_vuart> and <communication_vuart>, vuart_idx for PCI-vUART
 - 'base' in <console_vuart> and <communication_vuart>, PCI_VUART means use this PCI-vUART, it is a console. INVALID_PCI_BASE means do not use this PCI-VUART.
 - 'target_vm_id' and 'target_uart_id', connection settings for this vUART port.

> Run ``make BOARD_FILE=$PWD/misc/acrn-config/xmls/board-xmls/<board>.xml SCENARIO_FILE=/home/taoyuhong/acrn/acrn-hypervisor/misc/acrn-config/xmls/config-xmls/<board>/<scenario.xml``

Reviewer: ```suggestion
Run ``make BOARD_FILE=$PWD/misc/acrn-config/xmls/board-xmls/<board>.xml SCENARIO_FILE=$PWD/misc/acrn-config/xmls/config-xmls/<board>/<scenario.xml>``
```

Author: Will fix that

> to build ACRN with xml configuration file. Configuration tools will test your settings, check `vUART rules <https://projectacrn.github.io/latest/tutorials/vuart_configuration.html#index-of-vuart>`_ for compile issue. After compile, you can find ``./misc/vm_configs/scenarios/<scenario>/<board>/pci_dev.c`` has been changed by configuration tools, something like:

Reviewer: Replace with a :ref:`lavel`

Author: OK

.. code-block:: none

   {
      .emu_type = PCI_DEV_TYPE_HVEMUL,
      .vbdf.bits = {.b = 0x00U, .d = 0x05U, .f = 0x00U},
      .vdev_ops = &vmcs9900_ops,
      .vbar_base[0] = 0x80003000,
      .vbar_base[1] = 0x80004000,
      .vuart_idx = 1,
      .t_vuart.vm_id = 1U,
      .t_vuart.vuart_id = 1U,
    },

> That shows a PCI-vUART with vuart_idx=1, BDF 00:05.0, is PCI-vUART1 of this VM. We can use BDF to identify PCI-vUART in VM. Usually, legacy vuart[0] is ttyS0 in VM, and vuart[1] is ttyS1. So we hope PCI-vUART0 is ttyS0, PCI-VUART1 is ttyS1 ... PCI-vUART7 is ttyS7, but that is not true. If run ``dmesg | grep tty`` at VM shell, user may see:

Reviewer: Not sure what we are trying to say here. Can you explain some more?

Author: Because user need to know which `"/dev/ttyS*" is the right file, should be read/wrote for communication. For example, user know VM0 has PCI-vUART1 connected to VM1 PCI-vUART1, VM0 has PCI-vUART2 connected to VM2 PCI-vUART1. When VM0 want to talk to VM1, and VM0 has ttyS1, ttyS2 ... ttyS5 . In VM0, user may want to know which ttyS* is PCI-vUART1. It is not ttyS1 when using ACRN kernel. 
HV configuration tool assigns the BDF value, mmio address of VM0 PCI-vUART1. By checking BDF and mmio, user knows the mapping between "/dev/ttyS*" and vUART*

Reviewer: OK, thanks for the explanation, I think I understand this better now. We need to indicate that the use needs to retrieve which `ttyS*` is mapped to the device by inspecting `dmesg` to see which maps to the correct MMIO or BDF. Is that correct?

.. code-block:: none

   [    1.276891] 0000:00:05.0: ttyS4 at MMIO 0xa1414000 (irq = 124, base_baud = 115200) is a 16550A

So we know for guest OS, ttyS4 has BDF 00:05.0, is PCI-vUART1.

> if VM0 & VM1 are pre-launched VM, or SOS VM, ACRN hypervisor will create PCI-vUART virtual devices automatically. For post-launched VM, which is created by acrn-dm, additional acrn-dm option is needed to create PCI-vUART virtual device:

Reviewer: ```suggestion
if VM0 & VM1 are pre-launched VM, or Service VM, ACRN hypervisor will create PCI-vUART virtual devices automatically. For post-launched VM, which is created by ``acrn-dm``, additional ``acrn-dm`` option is needed to create PCI-vUART virtual device:
```

Author: Will fix that

.. code-block:: none

   -s <slot>,uart,vuart_idx:<val>

Kernel Config for Legacy vUART
==============================

> When ACRN hypervusor passthough local APIC to a VM, there is IRQ injection issue for legacy vUART. The kernel driver should work in polling mode to avoid the problem.

Reviewer: ```suggestion
When ACRN hypervisor passthrough local APIC to a VM, there is IRQ injection issue for legacy vUART. The kernel driver should work in polling mode to avoid the problem.
```

Author: Will fix it
